lib/segment/src/spaces/metric_f16/simple.rs

```diff
@@ -0,0 +1,201 @@
+use common::types::ScoreType;
+use half::f16;
+use num_traits::Float;
+
+#[cfg(target_arch = "x86_64")]
+use super::simple_avx::*;
+#[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+use super::simple_neon::*;
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+use super::simple_sse::*;
+use crate::data_types::vectors::{DenseVector, VectorElementTypeHalf};
+use crate::spaces::metric::{Metric, MetricPostProcessing};
+use crate::spaces::tools::is_length_zero_or_normalized;
+use crate::types::Distance;
+
+#[cfg(target_arch = "x86_64")]
+pub(crate) const MIN_DIM_SIZE_AVX: usize = 32;
+
+#[cfg(any(
+    target_arch = "x86",
+    target_arch = "x86_64",
+    all(target_arch = "aarch64", target_feature = "neon")
+))]
+pub(crate) const MIN_DIM_SIZE_SIMD: usize = 16;
+
+#[derive(Clone)]
+pub struct DotProductMetric;
+
+#[derive(Clone)]
+pub struct CosineMetric;
+
+#[derive(Clone)]
+pub struct EuclidMetric;
+
+#[derive(Clone)]
+pub struct ManhattanMetric;
+
+impl Metric<VectorElementTypeHalf> for EuclidMetric {
+    fn distance() -> Distance {
+        Distance::Euclid
+    }
+
+    fn similarity(v1: &[VectorElementTypeHalf], v2: &[VectorElementTypeHalf]) -> ScoreType {
+        #[cfg(target_arch = "x86_64")]
+        {
+            if is_x86_feature_detected!("avx")
+                && is_x86_feature_detected!("fma")
+                && v1.len() >= MIN_DIM_SIZE_AVX
+            {
+                return unsafe { euclid_similarity_avx(v1, v2) };
+            }
+        }
+
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        {
+            if is_x86_feature_detected!("sse") && v1.len() >= MIN_DIM_SIZE_SIMD {
+                return unsafe { euclid_similarity_sse(v1, v2) };
+            }
+        }
+
+        #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+        {
+            if std::arch::is_aarch64_feature_detected!("neon") && v1.len() >= MIN_DIM_SIZE_SIMD {
```

Comment:
float 16-bit neon was introduced in `armv8.2`, for `armv7.*` and `armv8.0` this code will produce segfault because of calling of non-exists CPU instruction. We need to check if float 16 is supported on device

Reply:
the same for each metric

Reply:
Added a check in build.rs to account for this.